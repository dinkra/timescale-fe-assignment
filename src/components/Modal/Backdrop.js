import styled from 'styled-components';

const StyledBackdrop = styled.div`
  position: fixed;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.7);
`;

const Backdrop = ({ toggleVisibility }) => {
  const handleClick = () => {
    if (typeof toggleVisibility !== 'undefined' && toggleVisibility) {
      toggleVisibility();
    }
  };

  return <StyledBackdrop onClick={handleClick} />;
};

export default Backdrop;
