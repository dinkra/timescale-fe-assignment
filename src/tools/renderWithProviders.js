import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import theme from '../theme';

const AllTheProviders = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';

// override render method
export { customRender as render };
