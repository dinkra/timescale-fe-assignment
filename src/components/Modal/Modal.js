import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';
import close from '../../theme/svgs/close';

// components
import Backdrop from './Backdrop';

// styles
const StyledModal = styled.div`
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  z-index: 250;
  background-color: ${({ theme }) => theme.white};
  padding: 20px 28px 28px;
  min-width: 360px;
`;

const StyledClose = styled.div`
  position: absolute;
  top: 20px;
  right: 20px;
  cursor: pointer;
`;

const Modal = ({ isVisible, toggleVisibility, children }) => {
  const onKeyDown = (event) => {
    if (event.key === 'Escape' && isVisible && toggleVisibility) {
      toggleVisibility();
    }
  };

  useEffect(() => {
    document.addEventListener('keydown', onKeyDown, false);

    if (isVisible) {
      document.body.style.setProperty('overflow', 'hidden');
    } else {
      document.body.style.removeProperty('overflow');
    }
    return () => {
      document.removeEventListener('keydown', onKeyDown, false);
      document.body.style.removeProperty('overflow');
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isVisible]);

  const modalElement = (
    <>
      <Backdrop toggleVisibility={toggleVisibility} />
      <StyledModal tabIndex={-1} role="dialog">
        <StyledClose data-testid="modal-close-btn" onClick={toggleVisibility}>
          {close}
        </StyledClose>
        {children}
      </StyledModal>
    </>
  );

  return isVisible ? ReactDOM.createPortal(modalElement, document.body) : null;
};

export default Modal;
