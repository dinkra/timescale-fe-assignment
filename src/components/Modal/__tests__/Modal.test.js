import React from 'react';
import { render, fireEvent, screen } from '../../../tools/renderWithProviders';

import Modal from '../';

describe('Modal', () => {
  const toggleVisibility = jest.fn();
  const onClose = jest.fn();

  test('should match snapshot', () => {
    const result = render(
      <Modal isVisible={true} toggleVisibility={toggleVisibility}>
        <div>Content</div>
      </Modal>
    );

    expect(result).toMatchSnapshot();
  });

  test('should fire close event', () => {
    const { getByText } = render(
      <Modal isVisible={true} toggleVisibility={toggleVisibility}>
        <div>Content</div>
      </Modal>
    );

    const closeBtn = screen.getByTestId('modal-close-btn');
    fireEvent.click(closeBtn);
    expect(toggleVisibility).toHaveBeenCalledTimes(1);
  });
});
