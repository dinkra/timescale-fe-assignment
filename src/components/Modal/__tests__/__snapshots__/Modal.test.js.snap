// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="overflow: hidden;"
  >
    <div />
    <div
      class="sc-bdnxRM czJYpv"
    />
    <div
      class="sc-gtsrHT eqkTRz"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="sc-dlnjwi eULyec"
        data-testid="modal-close-btn"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.71967 5.71967C6.01256 5.42678 6.48744 5.42678 6.78033 5.71967L12 10.9393L17.2197 5.71967C17.5126 5.42678 17.9874 5.42678 18.2803 5.71967C18.5732 6.01256 18.5732 6.48744 18.2803 6.78033L13.0607 12L18.2803 17.2197C18.5732 17.5126 18.5732 17.9874 18.2803 18.2803C17.9874 18.5732 17.5126 18.5732 17.2197 18.2803L12 13.0607L6.78033 18.2803C6.48744 18.5732 6.01256 18.5732 5.71967 18.2803C5.42678 17.9874 5.42678 17.5126 5.71967 17.2197L10.9393 12L5.71967 6.78033C5.42678 6.48744 5.42678 6.01256 5.71967 5.71967Z"
            fill="#141E35"
          />
          <rect
            height="23"
            stroke="#141E35"
            width="23"
            x="0.5"
            y="0.5"
          />
        </svg>
      </div>
      <div>
        Content
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
